spec/app/ics-020-fungible-token-transfer/README.md

```diff
@@ -210,33 +253,54 @@ In plain English, between chains `A` and `B`:
   an acknowledgement of failure is preferable to aborting the transaction since it more easily enables the sending chain
   to take appropriate action based on the nature of the failure.
 
+Note: `constructOnChainDenom` is a helper function that will construct the local on-chain denomination for the bridged token. It **must** encode the trace and base denomination to ensure that tokens coming over different paths are not treated as fungible. The original trace and denomination must be retrievable by the state machine so that they can be passed in their original forms when constructing a new IBC path for the bridged token. The ibc-go implementation handles this by creating a local denomination: `hash(trace+base_denom)`.
+
 `sendFungibleTokens` must be called by a transaction handler in the module which performs appropriate signature checks, specific to the account owner on the host state machine.
 
 ```typescript
 function sendFungibleTokens(
-  denomination: string,
-  amount: uint256,
+  tokens: []Token,
   sender: string,
   receiver: string,
+  memo: string,
   sourcePort: string,
   sourceChannel: string,
   timeoutHeight: Height,
   timeoutTimestamp: uint64): uint64 {
-    prefix = "{sourcePort}/{sourceChannel}/"
-    // we are the source if the denomination is not prefixed
-    source = denomination.slice(0, len(prefix)) !== prefix
-    if source {
-      // determine escrow account
-      escrowAccount = channelEscrowAddresses[sourceChannel]
-      // escrow source tokens (assumed to fail if balance insufficient)
-      bank.TransferCoins(sender, escrowAccount, denomination, amount)
-    } else {
-      // receiver is source chain, burn vouchers
-      bank.BurnCoins(sender, denomination, amount)
+    for token in tokens {
+      prefix = "{sourcePort}/{sourceChannel}/"
+      // we are the source if the denomination is not prefixed
+      source = token.trace[0] !== prefix
+      onChainDenom = constructOnChainDenom(token.trace, token.denom)
+      
+      if source {
+        // determine escrow account
+        escrowAccount = channelEscrowAddresses[sourceChannel]
+        // escrow source tokens (assumed to fail if balance insufficient)
+        bank.TransferCoins(sender, escrowAccount, onChainDenom, token.amount)
+      } else {
+        // receiver is source chain, burn vouchers
+        bank.BurnCoins(sender, onChainDenom, token.amount)
+      }
     }
 
-    // create FungibleTokenPacket data
-    data = FungibleTokenPacketData{denomination, amount, sender, receiver}
+    channel = provableStore.get(channelPath(sourcePort, sourceChannel))
+    if channel.version === "ics20-1" {
```

Comment:
related to the other point I made, we can't actually use channel version to determine which application version we're using. One issue came up. Let's say we have chainA and chainB, chain A only supports ics20-1, while chainB supports ics20-2. The channel opened is application version ics20-1.

Chain B receives a MsgTransfer which contains muiltiple denoms to transfer. In the case of ics20-1, we should reject the message unless the array contains a single element. 

Right now, I don't think we have a way to correctly determine which application version we currently are on as there is no additional state outside of the channel version.

Does it make sense to add an additional state entry specifying the application version within the channel module that can be used for this purpose? (if empty assume ics20-1)

Reply:
can access via ics4Wrapper, no need for additional state